### PR TITLE
Creating Time Base component for Media Contribution

### LINKED ISSUE #43: MMM Example Notebook: Time Varying Coefficients
To prove the flexibility and potential of framework (namely, bayesian stats and `pymc`), we would like to have an example notebook to illustrate how to extend the base model (based on simulated data) introduced in https://github.com/pymc-labs/pymmmc/pull/41 by allowing time varying coefficients via gaussian processes.

### PR DESCRIPTION
**Pull request by cetagostini**
## Description

Currently our model has a time-varying intercept within the `tvp.py` module. This module can be extended by adding a few simple changes.

The current implementation create a `time varying variable` to be applied on top of any other component in the model.

## Related Issue

- [x] Closes #43 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [x] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--752.org.readthedocs.build/en/752/

### CODE REVIEW

**Comment by cetagostini** [comment]
Current state:
- [x] Time varying intercept is working.
- [x] MMM Notebook works properly.

Next steps:
- [ ] Create a notebook to show case the relevance of consider time into the MMM.
- [ ] Test suite

File: pymc_marketing/mmm/tvp.py
Comment (wd60622) [inline pymc_marketing/mmm/tvp.py:107]: Might be better to just pass a subdictionary rather than having to add a name which creates a prefix. 

i.e. pass `model_config["intercept_tvp_config"]` instead of `model_config`
```python
# Example config
model_config = {
    "intercept_tvp_config": {
        "m": 200,
        "L": None,
        "eta_lam": 1,
        "ls_mu": None,
        "ls_sigma": 10,
        "cov_func": None,
    },
}
```

**Review by juanitorduz — CHANGES_REQUESTED** [review]
Loogre that our API changes are making these changes very easy 😎 ! Amazing!

I left some documentation comments. Mainly specifying we are using a global media GP multiplier rather than one GP factor per channel right?

**Comment by juanitorduz** [comment]
I also added this PR to the  `0.7.0` milestone as we seem very close, right?

**Comment by cetagostini** [comment]
> I also added this PR to the `0.7.0` milestone as we seem very close, right?

Yes, should be very easy! I'll go over the comments today/tomorrow, I'm sure we can merge by Friday!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ulfaslak) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:385]: Indeed. There should just be one `time_index` as before.

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (ulfaslak) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:407]: yeah this is valid. Right now `time_varying_prior` is structured such that it is centered on 1. Logic for that is:

```Python
phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
hsgp_coefs = pm.Normal(f"{name}_hsgp_coefs", dims=hsgp_dims)
f = phi @ (hsgp_coefs * sqrt_psd).T
centered_f = f - f.mean(axis=0) + 1  # <-- this line
```

Due to the scale of `f`, it's unlikely that it goes below 0, but it can happen (by a very small amount).

Could we floor it to zero?

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:110]: Good idea

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:407]: I thought there was a transformation on the f to make positive?

File: tests/mmm/test_lift_test.py
Comment (juanitorduz) [inline tests/mmm/test_lift_test.py:463]: What is the new "fast eval" ? 😎 ?

Comment (wd60622) [reply tests/mmm/test_lift_test.py:463]: Just this imported from Pymc

```python
def fast_eval(var):
    return function([], var, mode="FAST_COMPILE")()
```

Comment (juanitorduz) [reply tests/mmm/test_lift_test.py:463]: wow! I did not know this exit 🙌 . Thanks!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:110]: Just to be sure we are align, it's not a baseline. The baseline is the variable `channel_contributions`, the media varying variable would be more a time **scaling factor** adding variance to the mean.

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:385]: Good I added the change!

File: pymc_marketing/mmm/tvp.py
Comment (cetagostini) [reply pymc_marketing/mmm/tvp.py:107]: Done!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:392]: I'm good with both, I didn't made it but I can if you feel is better!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:398]: Related to the same comment the media multiplier is not the baseline, is more a scaling factor! But let's agree on the words 👀

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:541]: Same mentioned than before!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:406]: Guys, I was playing with this and I think the solution is to allow the distribution of the HSGP coefficients to be arbitrary, with default a only positive distribution. Basically follow the same pattern that we currently allow to all other distributions from the model. We could easily add to the "**_tvp_config", the keys "dist" and"kwargs".

The current setup give this prior for the HSGP:
image

If we switch to:
```python
        gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
        phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
        hsgp_coefs = pm.Gamma(f"_hsgp_coefs_{name}", size=hsgp_size, alpha=3, beta=7)#, alpha=10, beta=20)
        f = phi @ (hsgp_coefs * sqrt_psd).T
        centered_f = f - f.mean(axis=0) + 1
        return pm.Deterministic(name, centered_f, dims=dims)
```
The prior looks like:
image

The same happens if we play with a `Beta`
```python
        gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
        phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
        hsgp_coefs = pm.Beta(f"_hsgp_coefs_{name}", size=hsgp_size, alpha=3, beta=7)#, alpha=10, beta=20)
        f = phi @ (hsgp_coefs * sqrt_psd).T
        centered_f = f - f.mean(axis=0) + 1
        return pm.Deterministic(name, centered_f, dims=dims)
```
image

But everything looks very strange, if we force it to positive with `exp` for example.

image


I am not an expert in HSGPs but from a practical point of view it could be the answer. Maybe @bwengals , @ulfaslak , @juanitorduz who are more involved with HSGPs have the answer?

Does it sound reasonable to be able to allow another distribution that is not Normal? Or it should be normal for some specific property.

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py]: I really like this idea, and I think you're right. However, I think we could proceed with the first implementation and take some time in the next release to give this module more shape.

Why? My idea is that at some point, in the same way that users can now define which variable is hierarchical or not, or centered hierarchical or not, they can define which variable will be time-dependent, and surely that would require changing and giving it greater flexibility to the tvp module.

Currently we only allow two variables to be time dependent, ideally the user could decide which one varies. The planned release `0.8.0` is pointing in that direction and I think it would be the ideal place to spend some time thinking about and adding this appropriately!

For now, perhaps this allows us to have user insights, and begin to show the difference between having or not having time components!

**Comment by cetagostini** [comment]
Guys all the comments added, just waiting for feedback on how to move to guarantee a positive TVP in "medium". If you have any other comments about name changes or similar, you can make the direct commit and we will shorten the implementation time!

cc: @juanitorduz @wd60622

**Comment by juanitorduz** [comment]
@cetagostini I had to push a change on the tvp notebook to fix the pre-commit! can you fix the conflicts? (Maybe remove it from the PR and add the changes again?) Apologies for the inconvenience

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (cetagostini) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:406]: Indeed just for simplicity, I added `softplus` but not from `pytensor`.

**Comment by cetagostini** [comment]
Hey team, I made the required changes.

I see the implementation ready to do the merge, some things that I think will be interesting in the future:
1. Allow users to decide which variables are time dependent or not. Basically, create the flexibility so that each user can decide which model parameter should have time variability.
2. Allow, as with distributions, to manage the dimensions of the HSGP, that way, users could create latent variables around Geos, Channels, or other granularity.

Maybe we can handle these things for the next release. Once we have this merge, I start with the notebook (In other PR) to have everything ready for Monday's release!

**Comment by cetagostini** [comment]
PS: If you feel that a name could be more explicit, I would say commit directly!

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (wd60622) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:391]: Think `baseline` makes a bit more sense!

File: pymc_marketing/mmm/tvp.py
Comment (wd60622) [inline pymc_marketing/mmm/tvp.py:200]: I've been importing this from pymc
```python
from pymc.distributions.shape_utils import Dims
```
```suggestion
    dims: Dims,
```

File: pymc_marketing/mmm/tvp.py
Comment (wd60622) [inline pymc_marketing/mmm/tvp.py:223]: Think it is a bit more readable to pull out model_config[f"{name}_tvp_config"] before the function and pass that rather than do the indexing from within

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (wd60622) [inline tests/mmm/test_delayed_saturated_mmm.py:1]: Any other things we can tests besides just intialization works? 😆

File: pymc_marketing/mmm/tvp.py
Comment (cetagostini) [reply pymc_marketing/mmm/tvp.py:97]: Not in pytensor. At least didn't find it

Comment (wd60622) [reply pymc_marketing/mmm/tvp.py:97]: Seems to be there: https://github.com/search?q=repo%3Apymc-devs%2Fpytensor+softplus&type=code

```python
import pytensor.tensor as pt
pt.softplus
```

File: pymc_marketing/mmm/delayed_saturated_mmm.py
Comment (bwengals) [reply pymc_marketing/mmm/delayed_saturated_mmm.py:406]: @cetagostini I'd be super cautious about changing the distribution of the HSGP coefficients.  The reason their normal is for the "G" in "GP".  I've seen a paper where they use a student-t process and then use a student-t distribution here instead.  The standard way of forcing it positive is to pass it through `pt.exp`, but I suppose softplus would work too.   

If you want to also force it to be zero mean, you can subtract the mean of each basis vector from `phi` directly instead of doing `centered_f = f - f.mean(axis=0)`, which I think would work better.  Either of these will screw up forecasting though, which is the advantage of a GP here over splines or something like that.  I think the best option would be to either:
1. Get rid of the intercept parameter in the model, and just let the HSGP handle it.  If you're transforming it with `pt.exp` it should be able to handle some offsets.  This wouldn't work well if the scales of the intercept and GP were way different.  
2. Keep the intercept but set `drop_first` parameter in HSGP to True.  Depending on your `c` parameter, the first basis vector in `phi` can get very flat and intercept-like.

File: pymc_marketing/mmm/tvp.py
Comment (cetagostini) [reply pymc_marketing/mmm/tvp.py:97]: Done!

File: pymc_marketing/mmm/tvp.py
Comment (cetagostini) [reply pymc_marketing/mmm/tvp.py]: Done!

File: pymc_marketing/mmm/tvp.py
Comment (cetagostini) [reply pymc_marketing/mmm/tvp.py:207]: Done!

File: tests/mmm/test_delayed_saturated_mmm.py
Comment (cetagostini) [reply tests/mmm/test_delayed_saturated_mmm.py:1]: Quite right, I might need to think about it a bit more, I could try to add a few more but can plan a deep-dive for a next PR? ✌🏻

**Review by wd60622 — APPROVED** [review]
Thanks for making the changes! Don't love the idea of not fully testing. For instance, we dont know if there is a successful build with this model! However, I will trust you had some luck with it

**Comment by ulfaslak** [comment]
I think this all looks really good. May I have a stab at:

> Allow users to decide which variables are time dependent or not. Basically, create the flexibility so that each user can decide which model parameter should have time variability.

once we merge?

**Comment by wd60622** [comment]
> I think this all looks really good. May I have a stab at:
> 
> 
> 
> > Allow users to decide which variables are time dependent or not. Basically, create the flexibility so that each user can decide which model parameter should have time variability.
> 
> 
> 
> once we merge? My thought is that you should be able to pass a `bool` or a `list[str]` (list of channel names) to `time_varying_media`, and the latter would set gp multiplier on each listed channel.

Interesting. Thatd be cool. Might be related to #712

**Comment by juanitorduz** [comment]
> Thanks for making the changes! Don't love the idea of not fully testing. For instance, we dont know if there is a successful build with this model! However, I will trust you had some luck with it

Let's add a test that ensures this model builds and then we can merge :)

**Comment by cetagostini** [comment]
> > Thanks for making the changes! Don't love the idea of not fully testing. For instance, we dont know if there is a successful build with this model! However, I will trust you had some luck with it
> 
> Let's add a test that ensures this model builds and then we can merge :)

For sure, those test already exist on the `test_delayed_saturated` suite. You can check we validate the model is able to run with all other parameters on/off. Assuring it's compatible with other and it not breaks by combining them.

**Review by juanitorduz — APPROVED** [review]

**Pull request merged**
